src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,571 @@
+        self.start_callback_server(
+            docker.from_env(), uuid.uuid4().hex, job_dir, kb_base_url, token, port
+        )
+        ws_url = os.path.join(kb_base_url, "ws")
+        callback_url = "http://" + loader_helper.get_ip() + ":" + str(port)
+        print("callback_url:", callback_url)
+        self.ws = Workspace(ws_url, token=token)
+        self.asu = AssemblyUtil(callback_url, token=token)
+
+        self.workers = workers
+        self.input_queue = Queue()
+        self.output_queue = Queue()
+        self.pools = Pool(workers, process_input, [self])
+
+    def setup_callback_server_envs(self, job_dir, kb_base_url, token, port):
+        # initiate env and vol
+        env = {}
+        vol = {}
+
+        # used by the callback server
+        env["KB_AUTH_TOKEN"] = token
+        env["KB_BASE_URL"] = kb_base_url
+        env["JOB_DIR"] = job_dir
+        env["CALLBACK_PORT"] = port
+
+        # setup volumes required for docker container
+        docker_host = os.environ["DOCKER_HOST"]
+        if docker_host.startswith("unix:"):
+            docker_host = docker_host[5:]
+
+        vol[job_dir] = {"bind": job_dir, "mode": "rw"}
+        vol[docker_host] = {"bind": "/run/docker.sock", "mode": "rw"}
+
+        return env, vol
+
+    def start_callback_server(
+        self, client, container_name, job_dir, kb_base_url, token, port
+    ):
+        env, vol = self.setup_callback_server_envs(job_dir, kb_base_url, token, port)
+        self.container = client.containers.run(
+            name=container_name,
+            image=loader_common_names.CALLBACK_UPLOADER_IMAGE_NAME,
+            detach=True,
+            network_mode="host",
+            environment=env,
+            volumes=vol,
+        )
+        time.sleep(2)
+
+    def stop_callback_server(self):
+        self.container.stop()
+        self.container.remove()
+
+
+def _get_parser():
+    parser = argparse.ArgumentParser(
+        description="PROTOTYPE - Upload assembly files to the workspace service (WSS).",
+        formatter_class=loader_helper.ExplicitDefaultsHelpFormatter,
+    )
+
+    required = parser.add_argument_group("required named arguments")
+    optional = parser.add_argument_group("optional arguments")
+
+    # Required flag argument
+    required.add_argument(
+        "--workspace_id",
+        required=True,
+        type=int,
+        help="Target workspace addressed by the permanent ID",
+    )
+    required.add_argument(
+        f"--{loader_common_names.KBASE_COLLECTION_ARG_NAME}",
+        type=str,
+        help="The name of the collection being processed",
+    )
+    required.add_argument(
+        f"--{loader_common_names.SOURCE_VER_ARG_NAME}",
+        type=str,
+        help="The source version of the collection being processed",
+    )
+
+    # Optional argument
+    optional.add_argument(
+        "--root_dir",
+        type=str,
+        default=loader_common_names.ROOT_DIR,
+        help="Root directory.",
```

Comment:
The user needs more info than this - maybe look at some of the other tools, pick & improve the best description of the root dir, and make a common variable in the common loader names

Reply:
Updated globally!

Reply:
:+1: